cranelift/codegen/src/machinst/reg.rs

```diff
@@ -0,0 +1,432 @@
+//! Definitions for registers, operands, etc. Provides a thin
+//! interface over the register allocator so that we can more easily
+//! swap it out or shim it when necessary.
+
+use crate::machinst::MachInst;
+use alloc::{string::String, vec::Vec};
+use core::{fmt::Debug, hash::Hash};
+use regalloc2::{Allocation, Operand, PReg, VReg};
+use smallvec::{smallvec, SmallVec};
+
+/// The first 128 vregs (64 int, 64 float/vec) are "pinned" to
+/// physical registeres: this means that they are always constrained
+/// to the corresponding register at all use/mod/def sites.
+///
+/// Arbitrary vregs can also be constrained to physical registers at
+/// particular use/def/mod sites, and this is preferable; but pinned
+/// vregs allow us to migrate code that has been written using
+/// RealRegs directly.
+pub const PINNED_VREGS: usize = 128;
+
+/// Convert a `VReg` to its pinned `PReg`, if any.
+pub fn pinned_vreg_to_preg(vreg: VReg) -> Option<PReg> {
+    if vreg.vreg() < PINNED_VREGS {
+        Some(PReg::from_index(vreg.vreg()))
+    } else {
+        None
+    }
+}
+
+/// A register named in an instruction. This register can be either a
+/// virtual register or a fixed physical register. It does not have
+/// any constraints applied to it: those can be added later in
+/// `MachInst::get_operands()` when the `Reg`s are converted to
+/// `Operand`s.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Reg(VReg);
+
+impl Reg {
+    /// Extract the underlying `regalloc2::VReg`. Note that physical
+    /// registers also map to particular (special) VRegs, so this
+    /// method can be used either on virtual or physical `Reg`s.
+    pub fn to_vreg(self) -> VReg {
+        self.0
+    }
+
+    /// Create a `Reg` from an underlying `regalloc2::VReg`.
+    pub fn from_vreg(vreg: VReg) -> Self {
+        Self(vreg)
+    }
+
```

Comment:
`impl From<VReg> for Reg` and `impl From<Reg> for VReg`? This is a little bit more idiomatic and allows using it in generics, plus you get the corresponding `Into` blanket impl.

Reply:
Done!